modules/network/src/main/java/org/apache/ignite/network/NettyWorkersRegistrar.java

```diff
@@ -0,0 +1,147 @@
+     * of the worker corresponding to an event loop. If an event loop's thread is hung, the hearbeat will not
+     * be updated, and the worker watchdog will treat the event loop worker as blocked.
+     */
+
+    private final CriticalWorkerRegistry criticalWorkerRegistry;
+
+    private final ScheduledExecutorService scheduler;
+
+    private final NettyBootstrapFactory bootstrapFactory;
+
+    private final List<NettyWorker> workers = new CopyOnWriteArrayList<>();
+
+    @Nullable
+    private volatile ScheduledFuture<?> sendHearbeatsTaskFuture;
+
+    /**
+     * Constructor.
+     *
+     * @param criticalWorkerRegistry Used to register workers representing Netty event loops to detect when their threads are
+     *         blocked.
+     * @param scheduler Used to schedule periodic tasks.
+     * @param bootstrapFactory Used to obtain Netty workers.
+     */
+    public NettyWorkersRegistrar(
+            CriticalWorkerRegistry criticalWorkerRegistry,
+            ScheduledExecutorService scheduler,
+            NettyBootstrapFactory bootstrapFactory
+    ) {
+        this.criticalWorkerRegistry = criticalWorkerRegistry;
+        this.scheduler = scheduler;
+        this.bootstrapFactory = bootstrapFactory;
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        for (NioEventLoopGroup group : bootstrapFactory.eventLoopGroups()) {
+            registerWorkersFor(group);
+        }
+
+        sendHearbeatsTaskFuture = scheduler.scheduleAtFixedRate(this::sendHearbeats, 100, 100, MILLISECONDS);
+
+        return nullCompletedFuture();
+    }
+
+    private void registerWorkersFor(NioEventLoopGroup group) {
+        List<NettyWorker> groupWorkers = new ArrayList<>(group.executorCount());
+
+        for (EventExecutor eventExecutor : group) {
+            SingleThreadEventExecutor executor = (SingleThreadEventExecutor) eventExecutor;
+            groupWorkers.add(new NettyWorker(executor));
+        }
+
+        for (NettyWorker worker : groupWorkers) {
+            criticalWorkerRegistry.register(worker);
+        }
+
+        workers.addAll(groupWorkers);
+    }
+
+    private void sendHearbeats() {
+        for (NettyWorker worker : workers) {
+            worker.sendHeartbeat();
+        }
+    }
+
+    @Override
+    public void stop() throws Exception {
+        Future<?> heartBeatsTaskFuture = sendHearbeatsTaskFuture;
+        if (heartBeatsTaskFuture != null) {
+            heartBeatsTaskFuture.cancel(false);
+        }
+
+        for (NettyWorker worker : workers) {
+            criticalWorkerRegistry.unregister(worker);
+        }
+    }
+
+    private static class NettyWorker implements CriticalWorker {
+        private final SingleThreadEventExecutor eventExecutor;
+
+        private volatile long heartbeatNanos = System.nanoTime();
+
+        private final Runnable sendHeartbeatTask = () -> heartbeatNanos = System.nanoTime();
```

Comment:
Can it simply be a method?

Reply:
The idea was to spare an allocation of a lambda/method reference

Reply:
Is there a difference if you allocate it here or once inside `sendHeartbeat`?

Reply:
I suppose it will be allocated on each `sendHeartbeat()` call

Reply:
Having read on this, I'm not sure now. But I remember that a couple of times this (cache/not cache instances) my intuition betrayed me, so I still like the most obvious approach: no doubts can arise about whether it allocates a new instance each time or not :) We lose lazyness, but we already know that this will be called more than 0 times, so it's not a problem.